inc/Engine/CriticalPath/RESTWP.php

```diff
@@ -0,0 +1,480 @@
+
+/**
+ * Class RESTWP
+ *
+ * @package WP_Rocket\Engine\CriticalPath
+ */
+abstract class RESTWP {
+
+	/**
+	 * Namespace for REST Route.
+	 */
+	const ROUTE_NAMESPACE = 'wp-rocket/v1';
+
+	/**
+	 * Part of route namespace for this inherited class item type.
+	 *
+	 * @var string $route_namespace to be set with like post, term.
+	 */
+	protected $route_namespace;
+
+	/**
+	 * Responsible for dealing with data/database.
+	 *
+	 * @var DataManager datamanager instance.
+	 */
+	private $data_manager;
+
+	/**
+	 * Responsible for dealing with CPCSS APIs.
+	 *
+	 * @var APIClient api_client instance.
+	 */
+	private $api_client;
+
+	/**
+	 * RESTWP constructor.
+	 *
+	 * @since 3.6
+	 *
+	 * @param DataManager $data_manager Data manager instance, responsible for dealing with data/database.
+	 * @param APIClient   $api_client API Client instance to deal with CPCSS APIs.
+	 */
+	public function __construct( DataManager $data_manager, APIClient $api_client ) {
+		$this->data_manager = $data_manager;
+		$this->api_client   = $api_client;
+	}
+
+	/**
+	 * Registers the generate route in the WP REST API
+	 *
+	 * @since 3.6
+	 *
+	 * @return void
+	 */
+	public function register_generate_route() {
+		register_rest_route(
+			self::ROUTE_NAMESPACE,
+			'cpcss/' . $this->route_namespace . '/(?P<id>[\d]+)',
+			[
+				'methods'             => 'POST',
+				'callback'            => [ $this, 'generate' ],
+				'permission_callback' => [ $this, 'check_permissions' ],
+			]
+		);
+	}
+
+	/**
+	 * Register Delete CPCSS route in the WP REST API.
+	 *
+	 * @since  3.6
+	 */
+	public function register_delete_route() {
+		register_rest_route(
+			self::ROUTE_NAMESPACE,
+			'cpcss/' . $this->route_namespace . '/(?P<id>[\d]+)',
+			[
+				'methods'             => 'DELETE',
+				'callback'            => [ $this, 'delete' ],
+				'permission_callback' => [ $this, 'check_permissions' ],
+			]
+		);
+	}
+
+	/**
+	 * Checks user's permissions. This is a callback registered to REST route's "permission_callback" parameter.
+	 *
+	 * @since 3.6
+	 *
+	 * @return bool true if the user has permission; else false.
+	 */
+	public function check_permissions() {
+		return current_user_can( 'rocket_regenerate_critical_css' );
+	}
+
+	/**
+	 * Generates the CPCSS for the requested post ID.
+	 *
+	 * @since 3.6
+	 *
+	 * @param  WP_REST_Request $request WP REST request response.
+	 * @return WP_REST_Response
+	 */
+	public function generate( WP_REST_Request $request ) {
+		$item_id = intval( $request['id'] );
+		$output  = null;
+
+		// validate item.
+		$validated = $this->validate_item_for_generate( $item_id );
+		if ( ! is_wp_error( $validated ) ) {
+			// get item url.
+			$item_url = $this->get_url( $item_id );
+			$timeout  = ( isset( $request['timeout'] ) && ! empty( $request['timeout'] ) );
+
+			$output = $this->process_generate( $item_url, $item_id, $timeout );
+		}else {
+			$output = $this->return_error( $validated );
+		}
+
+		return rest_ensure_response( $output );
```

Comment:
```suggestion
		if ( is_wp_error( $validated ) ) {
			return rest_ensure_response( $this->return_error( $validated ) );
		}

                // get item url.
                $item_url  = $this->get_url( $item_id );
	        $timeout  = ( isset( $request['timeout'] ) && ! empty( $request['timeout'] ) );
		$output   = $this->process_generate( $item_url, $item_id, $timeout );

		return rest_ensure_response( $output );
```

Return early when we can, most of the time an `else` is not needed. Here by inverting the conditional logic, we can return early when there is an error, and process the rest after the conditional if not.